File: src/algorithm/factors/depth.ts
import { MS_PER_DAY, FREQUENCY_WINDOW_DAYS, DEPTH_MULTIPLIERS, MAX_DEPTH_MULTIPLIER } from '../constants'
import type { Interaction } from '../../types'

/**
 * Depth factor (0-1): average depth multiplier of recent interactions,
 * normalized by the max multiplier.
 */
export function computeDepth(
  interactions: Interaction[],
  now: number = Date.now(),
): number {
  const windowStart = now - FREQUENCY_WINDOW_DAYS * MS_PER_DAY
  const recent = interactions.filter((i) => i.date >= windowStart)

  if (recent.length === 0) return 0

  const totalDepth = recent.reduce((sum, i) => sum + DEPTH_MULTIPLIERS[i.type], 0)
  const avgDepth = totalDepth / recent.length
  return Math.min(avgDepth / MAX_DEPTH_MULTIPLIER, 1)
}

File: functions/algorithm/factors/depth.py
from typing import Any, List

from algorithm.constants import (
    DEPTH_MULTIPLIERS,
    FREQUENCY_WINDOW_DAYS,
    MAX_DEPTH_MULTIPLIER,
    MS_PER_DAY,
)


def compute_depth(interactions: List[Any], now: int) -> float:
    """Depth factor (0-1): average depth multiplier of recent interactions."""
    window_start = now - FREQUENCY_WINDOW_DAYS * MS_PER_DAY
    recent = [i for i in interactions if i["date"] >= window_start]

    if not recent:
        return 0.0

    total_depth = sum(DEPTH_MULTIPLIERS.get(i["type"], 1) for i in recent)
    avg_depth = total_depth / len(recent)
    return min(avg_depth / MAX_DEPTH_MULTIPLIER, 1.0)
